src/v/cluster/producer_state.cc

```diff
@@ -187,6 +187,11 @@ result<request_ptr> requests::try_emplace(
     // All invariants satisfied, enqueue the request.
     _inflight_requests.emplace_back(
       ss::make_lw_shared<request>(first, last, current, result_promise_t{}));
+    // if there are more than max cached requests inflight start dropping the
+    // oldest of them
+    if (_inflight_requests.size() > requests_cached_max) {
+        _inflight_requests.pop_front();
+    }
```

Comment:
One odd thing about popping here I think is there is a gap in sequence numbers between _finished_requests and _inflight_requests

eg: _finished = [1, 2, 3, 4, 5]
_inflight = [6, 7, 8, 9, 10, 11, 12]

we drop 6, 7, the whole sequence looks something like

[1, 2, 3, 4, 5] <gap> [8, 9, 10, 11, 12].. 


Wondering we should only pop off  "_finished_requests + finished inflight requests (not yet promoted)" if the combined count > 5, starting with _finished_requests.  protocol guarantees unfinished inflight <=5

Reply:
Maybe we can chat about that tomorrow, really i think we should promote the request to finished when its outcome is set i.e. when we do `request->set_result()` and then ignore the one coming from apply

Reply:
yes, the promotion during apply() seems like a real headache.

Reply:
Actually i was thinking about that, producer will never want to know anything about `6,7` as it is only interested with latest 5 requests which is in the inflight list